### ISSUE TITLE
Generally use `int64_t` to index arrays of H3 objects rather than `int`?

### ISSUE THREAD

**Issue by ajfriend** [uber/h3#432]
(This is coming up in my work on #413, but I figured it'd make sense to break it out into a separate discussion.)

Since we changed `maxH3ToChildrenSize` to return an `int64_t` rather than an `int` in #409 ([change](https://github.com/uber/h3/pull/409/files#diff-23107c47e5e19901207dc7e0816eff70e9100a9dd29ecf1ae590eed5a584463aL422-L425)), should we also change a few downstream functions to also return `int64_t`? Or, more generally, should we change any variable that keeps track of a number of H3 objects to be `int64_t`, since such a number could, in theory, be larger than `int`?

For example, since the `maxUncompactSize` function basically sums up values from `maxH3ToChildrenSize`, shouldn't it also return type `int64_t`?

We might also change the `uncompact` signature from something like

```c
int uncompact(
    H3Index *compactedSet,
    int numHexes,
    H3Index *h3Set,
    int maxHexes,
    int res
)
```

to something like
```c
int uncompact(
    H3Index *compactedSet,
    int64_t numHexes,
    H3Index *h3Set,
    int64_t maxHexes,
    int res
)
```

Maybe it would even make sense to define something like `typedef int64_t H3Size`, and use `H3Size` to index any collection of H3 objects?

Curious to know what other folks think!

**Comment by dfellis** [comment]
`int64_t` makes sense to me for these functions; even if the value it returns is far beyond what could be stored in memory, that would simply warn the user to not do that, such as taking `res0Cells` and trying to `uncompact` at res 15 to get ALL THE HEXAGONS is gonna fail hard on any hardware that currently exists. :)

**Comment by isaacbrodsky** [comment]
I believe the usual type for this is ‘size_t’, which is platform dependent and 64 bit on 64 bit systems.

**Comment by dfellis** [comment]
Only problem with `size_t` is that you'd get errors on some systems and not others. It's not a big deal, but it might be confusing to someone to get errors only when the compile H3 on embedded CPUs, or the like.

**Comment by ajfriend** [comment]
> `int64_t` makes sense to me for these functions; even if the value it returns is far beyond what could be stored in memory, that would simply warn the user to not do that, such as taking `res0Cells` and trying to `uncompact` at res 15 to get ALL THE HEXAGONS is gonna fail hard on any hardware that currently exists. :)

With the iterator work I'm doing, I'm imagining potential use cases where memory may not even be an issue. A user might be working with iterator objects that have very large sizes (and would thus need to know the true size), but those iterators may never be fully written to memory or even fully iterated through.

> I believe the usual type for this is ‘size_t’, which is platform dependent and 64 bit on 64 bit systems.

One problem with `size_t` is that subtraction with `size_t`s will also return a `size_t`, and so will often lead to unexpected arithmetic bugs. I've looked around in a few places ([one example](https://stackoverflow.com/a/34642843)), and it seems like there's a good amount of advice towards using signed types unless you're absolutely sure the unsigned type won't cause issues. My takeaway is that `size_t` makes sense if you're iterating forward through a list, but if you're doing any operations like comparing set sizes, set differences, etc., then `int64_t` is a safer choice as it avoids the potential mistakes around `size_t`.

**Comment by isaacbrodsky** [comment]
I think we can safely assume any platform running H3 is capable of using 64 bit integers, so I think that `int64_t` is a reasonable dependency. I think this is fine for platforms with larger bit widths because the libary has essentially no situations where we want to return larger arrays - there are not that many cells in the grid system. In 32 bit systems it may be an issue because we will not be able to pass it directly into `calloc`, in that case it's OK because it's a number of elements. This will mean we should have tests for 32 bit platforms (and update our test suite to reflect that)

**Comment by ajfriend** [comment]
Closed by #413
